core/templates/domain/exploration/RuleInputTypeFactory.ts

```diff
@@ -0,0 +1,157 @@
+// Copyright 2020 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Factory to determine correct type of RuleInputs
+ */
+
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
+
+import { IRuleInput } from 'domain/exploration/RuleObjectFactory';
+import { INumberWithUnitsBackendDict } from
+  'domain/objects/NumberWithUnitsObjectFactory';
+import { IFractionDict } from
+  'domain/objects/FractionObjectFactory.ts';
+/* eslint-disable max-len */
+import { INote } from 'extensions/interactions/MusicNotesInput/directives/music-notes-input-rules.service';
+import { IGraphBackendDict } from 'extensions/interactions/GraphInput/directives/graph-detail.service';
+/* eslint-enable max-len */
+
+@Injectable({
+  providedIn: 'root'
+})
+export class RuleInputTypeFactory {
+  _isGraph(variable: Object): variable is IGraphBackendDict {
```

Comment:
Could you please explain how `variable is IGraphBackendDict` works?

Reply:
It tells the ts compiler that the `variable` is of type `IGraphBackendDict`

Reply:
Hmm, but how is it different from `variable: IGraphBackendDict`?

Reply:
Case 1
```typescript
func1(a: Object): GraphDict {
...
}
```
It means that func1 returns a object whose type is GraphDict

Case 2
```typescript
func1(a: Object): a is GraphDict {
...
}
```
It means that func1 returns a `boolean` and it thats `true` it means a is of type GraphDict

Reply:
Ok, thanks

Reply:
I would suggest to add comments for all this. In general, wherever you think a reviewer might have a question, add comments in advance.